File: process/python/train.py
import xgboost as xgb
from numpy import array as numpy_array
from sklearn.metrics import mean_squared_error
from process.python.data import prep_data_for_training
from process.python.method import run_xgboost
from process.python.method import run_linear_regression
from process.python.eval import run_eval, run_feature_importance, run_plot
from process.python.vis import plot_data


def train_bc_model(
    obs: list,
    fcst: list,
    covariants: dict,
    test_size: float = 0.2,
    method: str = "xgboost",
    cfg={
        "xgboost": {
            "objective": "reg:squarederror",
            "n_estimators": 100,
            "learning_rate": 0.1,
            "max_depth": 3,
        }
    }
):
    """
    Perform bias correction using specified machine learning method.

    Parameters
        obs : list
            List of observed values
        fcst : list
            List of forecast values to be corrected
        test_size : float, optional (default=0.2)
            Proportion of the dataset to include in the test split (0 to 1)
        random_state : int or None, optional (default=None)
            Random seed for reproducibility of the train-test split
        method : str, optional (default="xgboost")
            Machine learning method to use for bias correction
            Options: "xgboost", "linear_regression"
        cfg : dict, optional
            Configuration dictionary for the selected method
            Default contains XGBoost parameters:
                - objective: "reg:squarederror"
                - n_estimators: 100
                - learning_rate: 0.1
                - max_depth: 3

    Returns: dict
        Results containing the trained model and predictions

    Notes
        This function assumes the existence of helper functions:
        - prep_data(): Prepares and splits the data
        - run_xgboost(): Runs XGBoost model
        - run_linear_regression(): Runs linear regression model
        - run_eval(): Evaluates model predictions

    Examples
        >>> obs = [1, 2, 3, 4, 5]
        >>> fcst = [1.1, 2.2, 3.1, 4.2, 5.1]
        >>> results = start_bc(obs, fcst, method="xgboost", show_metrics=True)
    """

    training_data = prep_data_for_training(fcst, covariants, obs, test_size=test_size)

    if method == "xgboost":
        results = run_xgboost(
            training_data["x_train"], training_data["y_train"], training_data["x_test"], cfg["xgboost"]
        )
    if method == "linear_regression":
        results = run_linear_regression(
            training_data["x_train"], training_data["y_train"], training_data["x_test"]
        )

    metrics = run_eval(results["y_pred"], training_data["y_test"])
    run_plot(fcst, obs, training_data, results)
    feature_importance = run_feature_importance(
        training_data["x_train"], training_data["y_train"], training_data["x_names"])

    print("<><><><><><><><><><><><>")
    print("Training evaluation (Metrics):")
    print(metrics)
    print("Training evaluation (Feature importance):")
    print(feature_importance)
    print("<><><><><><><><><><><><>")

    return {
        "model": results["model"], 
        "metrics": metrics, 
        "scaler": training_data["scaler"], 
        "feature_importance": feature_importance}

File: process/r/train.R
#' Starts a bias correction procedure using either XGBoost or linear regression.
#'
#' This function trains a model to correct the bias between observed and forecasted values.
#' It splits the data into training and testing sets, trains the specified model,
#' evaluates its performance, and returns the trained model, predictions, and metrics.
#'
#' @param obs Numeric vector of observed values.
#' @param fcst Numeric vector of forecasted values.
#' @param test_size Proportion of data to use for testing (default: 0.2).
#' @param random_state Seed for random number generation (default: NULL).
#' @param method Model to use for bias correction ("xgboost" or "linear_regression", default: "xgboost").
#' @param cfg List containing model-specific configurations.
#'            For XGBoost:
#'              - objective: objective function (default: "reg:squarederror").
#'              - nrounds: number of boosting rounds (default: 100).
#'              - eta: learning rate (default: 0.1).
#'              - max_depth: maximum tree depth (default: 3).
#'            For linear regression:
#'              - control_method: method for controlling model fitting (default: "cv").
#'              - cv_folds: number of cross-validation folds (default: 5).
#' @param show_metrics Logical, if TRUE, prints the evaluation metrics (default: FALSE).
#'
#' @return A list containing:
#'   - model: The trained model object.
#'   - prd: Numeric vector of predicted values on the test set.
#'   - metrics: List of evaluation metrics.
#'
#' @examples
#' obs <- rnorm(100)
#' fcst <- obs + rnorm(100, mean = 0.5, sd = 0.2)
#' results <- start_bc(obs, fcst, test_size = 0.3, method = "xgboost")
#' print(results$metrics)
#'
#' results_lm <- start_bc(obs, fcst, test_size = 0.3, method = "linear_regression")
#' print(results_lm$metrics)
#'
#' results_seed <- start_bc(obs, fcst, test_size = 0.3, random_state=123, method = "xgboost")
#' print(results_seed$metrics)
#'
#' results_metrics <- start_bc(obs, fcst, test_size = 0.3, show_metrics = TRUE, method = "xgboost")
#'
#' results_cfg <- start_bc(obs, fcst, test_size = 0.3, method = "xgboost", cfg = list(xgboost = list(objective = "reg:squarederror", nrounds = 50, eta = 0.05, max_depth = 4)))
#' print(results_cfg$metrics)
#'
#' \dontrun{
#' # Example cfg for linear regression
#' results_lm_cfg <- start_bc(obs, fcst, test_size = 0.3, method = "linear_regression", cfg = list(linear_regression = list(control_method = "cv", cv_folds = 10)))
#' print(results_lm_cfg$metrics)
#' }
#'
#' @importFrom caret createDataPartition
#' @importFrom utils head
#'
#' @export

train_bc_model <- function(
    obs,              
    fcst,
    covariants,
    test_size = 0.2,
    method = "xgboost",
    cfg = list(
      xgboost = list(
        objective = "reg:squarederror",
        nrounds = 100,
        eta = 0.1,
        max_depth = 3
      ),
      linear_regression = list(
        control_method = "cv",
        cv_folds = 5  
      )
    )
) {
  
  training_data <- prep_data_for_training(fcst, covariants, obs, test_size=test_size)

  if (method == "xgboost") {
    results <- run_xgboost(
      training_data$x_train,
      training_data$y_train,
      training_data$x_test,
      cfg$xgboost
    )
  }
  
  if (method == "linear_regression") {
    results <- run_linear_regression(
      training_data$x_train,
      training_data$y_train,
      training_data$x_test,
      cfg$linear_regression
    )
  }
  
  metrics <- run_eval(results$y_pred, training_data$y_test)
  run_plot(fcst, obs, training_data, results)
  feature_importance = run_feature_importance(
    training_data$x_train, training_data$y_train, training_data$x_names)
  
  cat("<><><><><><><><><><><><>\n")
  cat("Training evaluation (Metrics):\n")
  print(metrics)
  cat("Training evaluation (Feature importance):\n")
  print(feature_importance)
  cat("<><><><><><><><><><><><>\n")

  
  return(
    list(
      model=results$model,
      metrics=metrics,
      scaler=training_data$scaler
    )
  )
  
}